### PR TITLE
feat: add Renegade X support

### PR DESCRIPTION
**Pull request by RattleSN4K3**
Adds support for Renegade X, using master server
Related issues: #629

> [!NOTE]  
> Setup and Response updated to reflect made changes during developing this PR.

### Query single server
Setup:
```js
GameDig.query({
  type: 'renegade10',
  address: '1.0.0.1',
  port: 7778
```

Example response
```json
{
  "name": "[PREFIX] Server Name",
  "map": "CNC-Field",
  "password": false,
  "raw": {
    "Name": "Server Name",
    "NamePrefix": "[PREFIX]",
    "Current Map": "CNC-Field",
    "Bots": 0,
    "Players": 12,
    "Game Version": "Open Beta 5.84.767",
    "Variables": {
      "Mine Limit": 24,
      "bSteamRequired": false,
      "bPrivateMessageTeamOnly": false,
      "bPassworded": false,
      "bAllowPrivateMessaging": true,
      "bRanked": true,
      "Game Type": 1,
      "Player Limit": 64,
      "Vehicle Limit": 20,
      "bAutoBalanceTeams": false,
      "Team Mode": 6,
      "bSpawnCrates": true,
      "CrateRespawnAfterPickup": 35,
      "Time Limit": 50
    },
    "Port": 7778,
    "IP": "1.0.0.1"
  },
  "version": "Open Beta 5.84.767",
  "maxplayers": 64,
  "numplayers": 12,
  "players": [],
  "bots": [],
  "queryPort": 7777,
  "connect": "1.0.0.1:7778",
  "ping": 0
}
```

### Query and receive master list

Setup:
```js
GameDig.query({
  type: 'renegade10master'
```

Example response
```json
{
  "name": "",
  "map": "",
  "password": false,
  "raw": {
    "servers": [
      {
        "name": "Server Name",
        "map": "CNC-Walls",
        "password": false,
        "raw": {
          "Name": "Server Name",
          "Current Map": "CNC-Walls",
          "Bots": 0,
          "Players": 0,
          "Game Version": "Open Beta 5.71.534",
          "Variables": {
            "Mine Limit": 30,
            "bSteamRequired": false,
            "bPrivateMessageTeamOnly": false,
            "bPassworded": false,
            "bAllowPrivateMessaging": true,
            "bRanked": true,
            "Game Type": 1,
            "Player Limit": 24,
            "Vehicle Limit": 11,
            "bAutoBalanceTeams": false,
            "Team Mode": 6,
            "bSpawnCrates": true,
            "CrateRespawnAfterPickup": 30,
            "Time Limit": 0
          },
          "Port": 7777,
          "IP": "1.0.0.1"
        },
        "version": "Open Beta 5.71.534",
        "maxplayers": 24,
        "numplayers": 0,
        "players": [],
        "bots": [],
        "queryPort": 0
      },
      {
        "name": "[PREFIX] Test Server",
        "map": "CNC-Uphill",
        "password": true,
        "raw": {
          "Name": "Test Server",
          "NamePrefix": "[PREFIX]",
          "Current Map": "CNC-Uphill",
          "Bots": 0,
          "Players": 0,
          "Game Version": "Open Beta 5.79.698",
          "Variables": {
            "Mine Limit": 24,
            "bSteamRequired": false,
            "bPrivateMessageTeamOnly": false,
            "bPassworded": true,
            "bAllowPrivateMessaging": true,
            "bRanked": true,
            "Game Type": 1,
            "Player Limit": 32,
            "Vehicle Limit": 10,
            "bAutoBalanceTeams": false,
            "Team Mode": 1,
            "bSpawnCrates": false,
            "CrateRespawnAfterPickup": 30,
            "Time Limit": 60
          },
          "Port": 7777,
          "IP": "1.0.0.2"
        },
        "version": "Open Beta 5.79.698",
        "maxplayers": 32,
        "numplayers": 0,
        "players": [],
        "bots": [],
        "queryPort": 0
      }
    ]
  },
  "version": "",
  "maxplayers": 0,
  "numplayers": 0,
  "players": [],
  "bots": [],
  "connect": "undefined:undefined",
  "ping": 0
}
```

### CODE REVIEW

**Comment by RattleSN4K3** [comment]
ID test fails but I don't think the ID should be forced as `renegade10`.

**Comment by CosminPerRam** [comment]
Unfortunately it needs to be that due to the current GID formatting rules.
We have gotten a bunch of negative feedback regarding it and will most likely be changed in the next major version (see #514 for more info), but for now we've got to hold onto the old system.

**Comment by xCausxn** [comment]
Even though `0` is falsy, I think we need to make sure `password` is either `true` or `false`

**Comment by RattleSN4K3** [comment]
> Even though `0` is falsy, I think we need to make sure `password` is either `true` or `false`

Good catch. I actually wanted to enforce `"0"`/`"1"` and ended up having 0/1. The downside of working on severel protocols in parallel and and not checking the Results output.

**Comment by RattleSN4K3** [comment]
I changed processing `password`. Due to some commit message, I wanted to rebase the PR, but since it was a draft, I see no problem with it (I generel would not prefer doing so).

PR is ready for review.

File: protocols/renegadex.js [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    state.map = serverInfo['Current Map']
+    state.password = variables.bPassworded
+
+    state.numplayers = numplayers
+    state.maxplayers = variables['Player Limit'] || 0
+
+    // due to master server not providing bot/player list, and standard result has no bot count, add list with dummy values
+    state.players = Array.from(new Array(numplayers).keys(), (i) => ({ name: `Player #${i + 1}`, raw: {} }))
+    state.bots = Array.from(new Array(numbots).keys(), (i) => ({ name: `Bot #${i + 1}`, raw: {} }))
+
```
Comment (CosminPerRam) [inline protocols/renegadex.js:217]: In v4 we had players::setNum, which just filled the players field with empty values if the server wouldn't provide them, but it has been removed in https://github.com/gamedig/node-gamedig/pull/389 as you couldn't really determine whether the server provided the values but are as they are presented or filled by that method, please remove these.

**Review by CosminPerRam — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -3,6 +3,7 @@
 ## 5.X.Y
 * Feat: Replaced `punycode` package usage with `url.domainToASCII` (#630).
 * Feat: World of Padman (2007) - Added support (#636)
+* Feat: Renegade X (2014) - Added support (#643)
```
Comment (CosminPerRam) [inline CHANGELOG.md:6]: You could also mention yourself here if you'd like (:

File: protocols/renegadex.js (hunk 1 above)
Comment (RattleSN4K3) [reply protocols/renegadex.js:217]: The information of how many bots are set/playing whould be left out in the standard result, as #389 was only adding the player count. But yeah okay, could be something to improve/extend.

Comment (RattleSN4K3) [reply protocols/renegadex.js]: Remove the lines. But also passed the raw server info as output now,

File: CHANGELOG.md (hunk 2 above)
Comment (RattleSN4K3) [reply CHANGELOG.md:7]: It's fine without

File: protocols/renegadex.js [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  constructor () {
+    super()
+    this.usedTcp = true
+  }
+
+  async run (state) {
+    const servers = await this.request({
+      url: 'https://serverlist-rx.totemarts.services/servers.jsp',
+      responseType: 'json'
+    })
```
Comment (CosminPerRam) [inline protocols/renegadex.js:170]: Please split this protocol to a master one (`renegadexmaster`) and this one (`renegadex`) to let users to query just the master server to obtain the server list if they'd like to.

Other similar cases of this: 
* [beammpmaster](https://github.com/gamedig/node-gamedig/blob/c77ca299b017619506bdb0680e5809e36364520f/protocols/beammpmaster.js) - [beammp](https://github.com/gamedig/node-gamedig/blob/c77ca299b017619506bdb0680e5809e36364520f/protocols/beammp.js)
* [vintagestorymaster](https://github.com/gamedig/node-gamedig/blob/c77ca299b017619506bdb0680e5809e36364520f/protocols/vintagestorymaster.js) - [vintagestory](https://github.com/gamedig/node-gamedig/blob/c77ca299b017619506bdb0680e5809e36364520f/protocols/vintagestory.js)
* [tribes1master](https://github.com/gamedig/node-gamedig/blob/c77ca299b017619506bdb0680e5809e36364520f/protocols/tribes1master.js) - note that [tribes1](https://github.com/gamedig/node-gamedig/blob/c77ca299b017619506bdb0680e5809e36364520f/protocols/tribes1.js) does not use the master server for the query, but we still have the master server implementation.

**Review by CosminPerRam — CHANGES_REQUESTED** [review]

File: protocols/renegadex.js (hunk 3 above)
Comment (RattleSN4K3) [reply protocols/renegadex.js:170]: That master-separation doesn't seem to be fully implemented (respectively exposed for the API). The master query logic is also a bit inconsistent throughout the protocols (see `altvml` and `ragemp` [or `minetest`]).

Also by having a master protocol (like `vintagestorymaster`) there could be the assumption of receiving a list of processed server data and not just the raw master server data. And vice versa

Not sure how I feel about that. Should I still follow the mentioned split-master-logic?

Comment (CosminPerRam) [reply protocols/renegadex.js:170]: I'd say yes, although we don't really mention anything about the support of querying master servers (and that it is flaky right now) we can provide it as just it is right now and straighten it up for the next major release (we should do a documented issue about this).

Comment (RattleSN4K3) [reply protocols/renegadex.js]: Added that master protocol. I can't tell if I like the current logic of adding master-protocols (the earlier subsystem I propsoed in #638 could be a alternative), but here we go.

The master server can be used wih `protocol-renegadexmaster` (or to be consistent to current game ID naming, `protocol-renegade10master`. The server list will be processed to not provide different type of data.

File: protocols/index.js [hunk 4]
```diff
@@ -37,6 +37,7 @@ import palworld from './palworld.js'
 import quake1 from './quake1.js'
 import quake2 from './quake2.js'
 import quake3 from './quake3.js'
+import renegadex, { renegadexmaster } from './renegadex.js'
```
Comment (CosminPerRam) [inline protocols/index.js:40]: Please split the master protocol query in another file.

File: protocols/index.js [hunk 5]
```diff
@@ -67,7 +68,7 @@ import soldat from './soldat.js'
 export {
   armagetron, ase, asa, assettocorsa, battlefield, buildandshoot, cs2d, discord, doom3, eco, epic, factorio, farmingsimulator, ffow,
   fivem, gamespy1, gamespy2, gamespy3, geneshift, goldsrc, gtasao, hexen2, jc2mp, kspdmp, mafia2mp, mafia2online, minecraft,
-  minecraftbedrock, minecraftvanilla, minetest, mumble, mumbleping, nadeo, openttd, palworld, quake1, quake2, quake3, rfactor, ragemp, samp,
+  minecraftbedrock, minecraftvanilla, minetest, mumble, mumbleping, nadeo, openttd, palworld, quake1, quake2, quake3, renegadex, renegadexmaster, renegadexmaster as renegade10master, rfactor, ragemp, samp,
```
Comment (CosminPerRam) [inline protocols/index.js:71]: Export only one entry, which should be named as `${game_id}master`, so `renegadexmaster`.
(side note: this is also not specified anywhere nor standalized unfortunately)

**Review by CosminPerRam — CHANGES_REQUESTED** [review]

File: protocols/index.js (hunk 5 above)
Comment (RattleSN4K3) [reply protocols/index.js:71]: That's the issue as the _game id_ is forced to be `...10`. `renegadex...` would be wrong to due to  the strict _normalisation_. Exporting the "normalized" name was just a shortcut solution.

File: protocols/index.js (hunk 4 above)
Comment (RattleSN4K3) [reply protocols/index.js:40]: I would have thought protocols (as file) provide the logic for the single protocol. Thus as both classes are using the master-server, I kept it in the same file.

With possible changes (for instance with #608) the file structured could be scattered.

**Comment by RattleSN4K3** [comment]
Updated the PR to reflect the proposed changes.

My points regarding the last changes still stand tho:
- Alternative protocols in a single file could be messy... in case of protocols being dependet on existing ones (see `master` and server-query based on `master` protocol)
  > With possible changes (for instance with #608) the file structured could be scattered.
- Inconsistency in protocol names and game id

**Review by CosminPerRam — APPROVED** [review]

**Pull request merged**